### PR TITLE
Add mongodb memory server

### PR DESCRIPTION
**Pull request by pdfowler**
I have a fair amount of experience with `mongo-memory-server` and after seeing the difficulties with PR #1078 I thought I'd see if I could get things working. 

## Highlights

1. By default tests will be executed against an in-memory instance of MongoDB
2. Tests can be executed against a hosted MongoDB instance by setting the `USE_MONGODB` env variable to a truthy value (yes|true|1).
3. Travis and Docker configurations will not use the new in-memory db instance.

### Misc
* Some fragile tests were "hardened"
* Logging adjustments
* Linting configuration specific to the `/test` directory

## Todo

- [x] Look into auto formatting abnormalities (ie: `test/retry.js` appears to have tabs in the diff)
- [ ] Set a fixed MongoDB version for `mongo-memory-server`

### CODE REVIEW

**Comment by Uzlopak** [comment]
I saw your PR already a week ago. Actually I ported for our needs agendajs to typescript and also use mongoose instead of direct mongo driver. So for us all is up and running and "fully" tested. Somehow agendajs is janky and I am kind of feeling that it is kind of unstable. So I am happy if there are better tests/code so that I can port it to our implementation.

**Comment by koresar** [comment]
As a hardcore node developer I'd like to clap to this PR. Brilliant!

Changes look minimalistic clean and thoroughly thought through.

I've pulled the branch, run the tests on my machine, - all good!

Let's merge it.

Unrelated: I tried replacing all `delay(50)` with `delay(5)` and the testing time dropped from 45 to 24 seconds.

**Review by koresar — DISMISSED** [review]

**Comment by simison** [comment]
@Uzlopak @pdfowler ah bummer, our typescript rewrite stepped on this PR just before merging.

Would you be so kind and follow-up on renamed `.ts` files and we can then merge this? Sorry for inconvenience and thanks for your work thus far!

**Comment by simison** [comment]
>    Some fragile tests were "hardened"
>    Logging adjustments
>    Linting configuration specific to the /test directory

These might better be in their own PRs so that it's easier to review/rebase/merge.

File: test/job.js
Comment (pdfowler) [inline test/job.js:22]: Seems that travis got dropped since my original work on this ... tagging line for removal

File: test/job.js
Comment (pdfowler) [inline test/job.js:51]: I'm not clear on the order of changes WRT the mongo connect options. Were these part of new tests/functionality or are they safe to ignore?

**Comment by pdfowler** [comment]
@simison I finally got a chance to dust off the code and get upstream/master merged in. After further review, I'm left with the following questions/issues/outstanding tasks:

[] Extracting out the non-core code from this PR (partially resolved w/ elimination of `xo`)
[] Determine whether the [mongo config options](https://github.com/agenda/agenda/pull/1108#pullrequestreview-692525434) are core to the tests (currently this PR removes them)
[] Decide on behavior of failing test: [job methods:unique:...:should modify one job when unique matches](https://github.com/agenda/agenda/blob/c29d2141a314429f2600bac184a26167ae1fcf8f/test/agenda.js#L416)

File: test/agenda.js
Comment (pdfowler) [inline test/agenda.js:416]: This test is failing on both this branch and master

File: test/job.js
Comment (koresar) [reply test/job.js]: Yes, please remove travis. Instead you might want to use the standard `CI` env var.

```
if (process.env.CI)
```

File: test/job.js
Comment (koresar) [reply test/job.js:50]: They are useful for old mongodb versions. I think prior to 3.6. But we do not support older mongodb any more. So, it's okay to drop these options.

**Review by koresar — CHANGES_REQUESTED** [review]
Just replace `process.env.TRAVIS` with the more universal `process.env.CI` and we are good to go!

File: test/job.js
Comment (pdfowler) [reply test/job.js]: @koresar Github PR diffs can be a bit strange at times: that line was still showing in the PR, but had been removed when I pulled it locally. I've pushed a change to restore it to the flat `const jobTimeout = 50;`

**Comment by pdfowler** [comment]
@koresar Just pushed the change to restore the `jobTimeout` variable to use the `process.env.CI` ([test/job.js](https://github.com/agenda/agenda/pull/1108/files?file-filters%5B%5D=.js&file-filters%5B%5D=.lock&file-filters%5B%5D=.yml#diff-3d864fcfcae7cc47fa275a1fac955f5a2e5f44538c11c65ea1342fbdd1717ed1R20)).

The only remaining question is WRT the failing "should modify one job when unique matches" test in `test/agenda.js`. I've [added a comment](https://github.com/agenda/agenda/pull/1108/files?file-filters%5B%5D=.js&file-filters%5B%5D=.lock&file-filters%5B%5D=.yml#diff-c5a2911b4cb8dd6e1674860f862942dbd34f7a91776c2e4b7de43728e5b26da2R415-R416) there pointing to a possible resolution, but wasn't sure it should be addressed as part of this PR or not.

File: test/agenda.js
Comment (koresar) [inline test/agenda.js:415]: Please remove this comment. It's not relevant any more. The unit test got broken with a different PR. The test is passing now. And the test is correct. It must make sure that the `nextRunAt` changed. Later on it checks that there is only one job in the DB.

**Review by koresar — CHANGES_REQUESTED** [review]

**Comment by koresar** [comment]
@pdfowler I've just sent you a request to become one of the maintainers. You can now create branches directly within Agenda and all its repos. I hope you'll find it useful.
Also, you can now close tickets and merge PRs.

**Comment by harisvsulaiman** [comment]
@pdfowler Welcome to the team. I hope we can transform agenda into the best job scheduler with good persistence gaurentees.

**Review by koresar — APPROVED** [review]

File: test/find-and-lock-next-job.js
Comment (harisvsulaiman) [inline test/find-and-lock-next-job.js:37]: There is no jobs here. Did you mean agenda? Related to https://github.com/agenda/agenda/issues/1369.

Comment (koresar) [reply test/find-and-lock-next-job.js:37]: That's an after merge bug. Could you please fix it in `master` @harisvsulaiman ?

**Pull request merged**
